pkg/cmd/server/admin/decrypt.go

```diff
@@ -0,0 +1,158 @@
+package admin
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/openshift/origin/pkg/cmd/util"
+	"github.com/spf13/cobra"
+
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
+	pemutil "github.com/openshift/origin/pkg/cmd/util/pem"
+)
+
+const DecryptCommandName = "decrypt"
+
+type DecryptOptions struct {
+	Data    *pem.Block
+	InFile  string
+	OutFile string
+
+	Key       []byte
+	KeyArg    string
+	InKeyFile string
+
+	Input  io.Reader
+	Output io.Writer
+	Stderr io.Writer
+}
+
+const decryptExample = `	# Decrypt using a key file, and output the decrypted value:
+	$ %[1]s --in=secret1.data --keyin=secret1.key
+	
+	# Decrypt using a key file, writing the decrypted value to a file:
+	$ %[1]s --in=secret2.data --keyin=secret1.key --out=secret2.txt
+
+	# Decrypt with a prompted password:
+	$ %[1]s --in=secret3.data
+`
+
+func NewCommandDecrypt(commandName string, fullName, encryptFullName string, out io.Writer) *cobra.Command {
+	options := &DecryptOptions{Output: out}
+
+	cmd := &cobra.Command{
+		Use:     commandName,
+		Short:   fmt.Sprintf("Decrypt data encrypted with %q", encryptFullName),
+		Example: fmt.Sprintf(decryptExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Validate(args))
+			kcmdutil.CheckErr(options.Complete(args))
+			kcmdutil.CheckErr(options.Decrypt())
+		},
+	}
+
+	flags := cmd.Flags()
+
+	flags.StringVar(&options.InFile, "in", options.InFile, fmt.Sprintf("File containing the data to decrypt, in the format written by %q.", encryptFullName))
+	flags.StringVar(&options.OutFile, "out", options.OutFile, "File to write the decrypted data to. Written to stdout if omitted.")
+
+	flags.StringVar(&options.KeyArg, "key", options.KeyArg, "The password to use to decrypt. Prompted for if omitted.")
+	flags.StringVar(&options.InKeyFile, "keyin", options.InKeyFile, fmt.Sprintf("The file to read the decrypting key from. Must be a PEM file in the format written by %q.", encryptFullName))
+
+	// autocompletion hints
+	cmd.MarkFlagFilename("in")
+	cmd.MarkFlagFilename("out")
+	cmd.MarkFlagFilename("keyin")
+
+	return cmd
+}
+
+func (o *DecryptOptions) Complete(args []string) error {
+	if o.Stderr == nil {
```

Comment:
Not sure I like this construct.  `Complete` is normally about taking your parameters and args and making the full object.  Not filling fields with "best effort" defaults.  I don't feel very strongly.